File: itops2/app/core/dates.py
"""Small date-arithmetic helpers with no external dependency
(no python-dateutil in requirements.txt)."""

from datetime import date


def add_months(d: date, months: int) -> date:
    """Calendar month arithmetic: rolls the year forward as needed and
    clamps the day into the target month (e.g. 31 Jan + 1 month ->
    28/29 Feb, not an invalid date)."""
    month_index = d.month - 1 + months
    year = d.year + month_index // 12
    month = month_index % 12 + 1
    day = min(
        d.day,
        [
            31,
            29 if year % 4 == 0 and (year % 100 != 0 or year % 400 == 0) else 28,
            31, 30, 31, 30, 31, 31, 30, 31, 30, 31,
        ][month - 1],
    )
    return date(year, month, day)

File: 001 - New project/coffer/lib/coffer/dates.ex
defmodule Coffer.Dates do
  @moduledoc """
  Small date-arithmetic helpers shared by anything that advances a date by a
  recurring cycle (contract renewals, recurring ledger transactions, ...).
  """

  @doc """
  Elixir's `Date` has no month arithmetic — adds `months` to `date`,
  clamping day-of-month overflow to the target month's last valid day
  (e.g. Jan 31 + 1 month = Feb 28/29, never an invalid "Feb 31").
  """
  def add_months(%Date{} = date, months) when is_integer(months) do
    total_months = date.year * 12 + (date.month - 1) + months
    year = div(total_months, 12)
    month = rem(total_months, 12) + 1
    last_day = Date.new!(year, month, 1) |> Date.days_in_month()
    Date.new!(year, month, min(date.day, last_day))
  end
end
